Improved ExtensiblePlugin

henningandersen opened the pull request:
Rather than let ExtensiblePlugins know extending plugins' classloaders,
we now pass along an explicit ExtensionLoader that loads the extensions
asked for. Extensions constructed that way can optionally receive their
own Plugin instance in the constructor.

A demonstration that this works for adding an ML decider for autoscaling is [here](https://github.com/elastic/elasticsearch/compare/master...henningandersen:enhance_autoscaling_plugability_beta?expand=1)

This also demonstrates one possible solution to [loading extensions in integration tests](
https://github.com/elastic/elasticsearch/compare/master...henningandersen:enhance_autoscaling_plugability_beta?expand=1#diff-7ce519dfb4c53aeaec1ff17fd46bd4b9R77)

but I would prefer to tackle that in a follow-up.

henningandersen committed:
Improved ExtensiblePlugin

Rather than let ExtensiblePlugins know extending plugins' classloaders,
we now pass along an explicit ExtensionLoader that loads the extensions
asked for. Extensions constructed that way can optionally receive their
own Plugin instance in the constructor.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Plugins)

henningandersen committed:
Fix Security.java

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -426,17 +429,66 @@ private static void addSortedBundle(Bundle bundle, Map<String, Bundle> bundles,
+    }
+
+    private static void loadExtensionsForPlugin(ExtensiblePlugin extensiblePlugin, List<Plugin> extendingPlugins) {
+        ExtensiblePlugin.ExtensionLoader extensionLoader = new ExtensiblePlugin.ExtensionLoader() {
```
This could just be created in loadExtensions, no need to create a new one for each individual plugin? AFAICT it always operates on the same stream of plugins passed in.

henningandersen replied:
The set of plugins passed in are only those extending the plugin that we load extensions for. This ensures that a plugin cannot ask for extensions that it does not own. Like ML asking for all autoscaling-deciders, which would succeed if the list of plugins was every plugin extending some other plugins. Also, the runtime would be slightly worse I think, since we would also be searching for extensions in unrelated modules.

rjernst replied:
Thanks for the explanation. I missed the lookup in calling of this method, thinking the extendingPlugins in the calling function (a map) was the same as extendingPlugins here (the value from that map, a list). Maybe consider changing the variable name to clarify they are different.

rjernst commented on server/src/main/java/org/elasticsearch/plugins/ExtensiblePlugin.java:
```diff
@@ -27,8 +29,22 @@
+         * @param <T> extension point type
+         * @return stream of all implementing extensions.
+         */
+        <T> Stream<T> loadExtensions(Class<T> extensionPointType);
```
Why a stream? It seems to lend itself to longer lines, but not easier to read code in this case, especially in the implementation below

henningandersen replied:
I changed it to a List. Stream is nice though, since its contract is read-only and read-once - and still easy to use (unlike Iterable and Iterator).

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -426,17 +429,66 @@ private static void addSortedBundle(Bundle bundle, Map<String, Bundle> bundles,
+        try {
+            try {
+                return extensionClass.getConstructor(plugin.getClass()).newInstance(plugin);
+            } catch (NoSuchMethodException e) {
```
I think we should lookup constructors just like in `loadPlugin`, rather than using exceptions as control flow here. The suppressed exception in the case we use the zero arg constructor would be especially confusing.

rjernst commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -509,12 +561,13 @@ private Plugin loadBundle(Bundle bundle, Map<String, Plugin> loaded) {
-        }
 
         Class<? extends Plugin> pluginClass = loadPluginClass(bundle.plugin.getClassname(), loader);
+        if (loader != pluginClass.getClassLoader()) {
```
When can this happen?

henningandersen replied:
It is not so much that it could happen, I hope not. But we rely on this in `createExtensions`:
https://github.com/elastic/elasticsearch/pull/58234/files#diff-40d98239fea26b07396b5032b95e396fR464

which is why I introduced this check which seems intuitively correct?

rjernst replied:
I understand now. Could we possibly have a test for this case?

henningandersen replied:
Turned out not too difficult, added a test here: eec9b68203ba9748543d2b4f5709dc19a2accc61

rjernst commented on server/src/test/java/org/elasticsearch/plugins/PluginsServiceTests.java:
```diff
@@ -690,4 +695,54 @@ public void testExistingMandatoryInstalledPlugin() throws IOException {
+        assertThat(extensiblePlugin.extensions.get(1), instanceOf(TestExtension2.class));
+        assertThat(((TestExtension2) extensiblePlugin.extensions.get(1)).plugin, sameInstance(testPlugin));
+    }
+
```
Can we have some tests for the error cases? For example, no valid ctor exists, or we fail when calling the ctor?

rjernst reviewed (changes requested):
This looks great! I have a few requests.

henningandersen committed:
Nicer errors on bad constructors.

henningandersen committed:
Verify exception in constructor.

henningandersen committed:
Remove some stream usages.

henningandersen commented on server/src/main/java/org/elasticsearch/plugins/PluginsService.java:
```diff
@@ -426,17 +429,66 @@ private static void addSortedBundle(Bundle bundle, Map<String, Bundle> bundles,
+        Map<String, List<Plugin>> extendingPlugins = plugins.stream()
+            .flatMap(t -> t.v1().getExtendedPlugins().stream().map(extendedPlugin -> Tuple.tuple(extendedPlugin, t.v2())))
+            .collect(Collectors.groupingBy(Tuple::v1, Collectors.mapping(Tuple::v2, Collectors.toList())));
+        plugins.stream().filter(t -> t.v2() instanceof ExtensiblePlugin).forEach(t -> loadExtensionsForPlugin((ExtensiblePlugin) t.v2(),
```
I changed this into a for loop instead, but were unable to use `filterPlugins`, since:

1. It operates on the full list of plugins, which includes test plugins.
2. It returns plugins, thus I loose the name of the module/bundle, which seems like the most natural key to use for the map above.

henningandersen commented:
Thanks for reviewing @rjernst , I believe I addressed all your comments so this should be ready for another round.

henningandersen commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhance_extensible_plugin

rjernst reviewed (approved):
LGTM

henningandersen committed:
Rename variable to avoid confusion

henningandersen committed:
Add test case

Added test case for defining a plugin using a Plugin class from the
parent class loader.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into enhance_extensible_plugin

henningandersen commented:
@elasticmachine run elasticsearch-ci/2
Test failure is new, fixed in: https://github.com/elastic/elasticsearch/pull/58525

henningandersen commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhance_extensible_plugin
